packages/orchestrator/internal/tcpfirewall/handlers.go

```diff
@@ -0,0 +1,182 @@
+	// This prevents goroutine leaks from slow/unresponsive DNS or connections.
+	upstreamDialTimeout = 30 * time.Second
+
+	noHostnameValue = ""
+)
+
+func domainHandler(ctx context.Context, conn net.Conn, dstIP net.IP, dstPort int, sbx *sandbox.Sandbox, logger logger.Logger) {
+	sourceAddr := conn.RemoteAddr().String()
+
+	// Get hostname from tcpproxy's wrapped connection
+	var hostname string
+	if tc, ok := conn.(*tcpproxy.Conn); ok {
+		hostname = tc.HostName
+	}
+
+	if hostname == "" {
+		// No hostname found, this is the case e.g. for https://1.1.1.1 like requests
+		logger.Debug(ctx, "No hostname found, ignoring hostname based filter", zap.String("source_addr", sourceAddr))
+	}
+
+	allowed, err := isEgressAllowed(sbx, hostname, dstIP)
+	if err != nil {
+		logger.Error(ctx, "Egress check failed", zap.Error(err))
+		conn.Close()
+
+		return
+	}
+
+	if !allowed {
+		logger.Debug(ctx, "Denied sandbox egress connection",
+			zap.String("hostname", hostname),
+			zap.String("destination_ip", dstIP.String()),
+			zap.String("source_addr", sourceAddr),
+		)
+		conn.Close()
+
+		return
+	}
+
+	// Proxy to IP directly (hostname already resolved by the client)
+	upstreamAddr := net.JoinHostPort(dstIP.String(), fmt.Sprintf("%d", dstPort))
+
+	logger.Debug(ctx, "Proxying sandbox egress hostname filter connection",
+		zap.String("hostname", hostname),
+		zap.String("source_addr", sourceAddr),
+		zap.String("upstream_addr", upstreamAddr),
+	)
+
+	proxy(conn, upstreamAddr)
+}
+
+func cidrOnlyHandler(ctx context.Context, conn net.Conn, dstIP net.IP, dstPort int, sbx *sandbox.Sandbox, logger logger.Logger) {
+	sourceAddr := conn.RemoteAddr().String()
+
+	// No hostname available for CIDR-only handler
+	allowed, err := isEgressAllowed(sbx, noHostnameValue, dstIP)
+	if err != nil {
+		logger.Error(ctx, "Egress check failed", zap.Error(err))
+		conn.Close()
+
+		return
+	}
+
+	if !allowed {
+		logger.Debug(ctx, "Denied sandbox egress connection",
+			zap.String("destination_ip", dstIP.String()),
+			zap.String("source_addr", sourceAddr),
+		)
+		conn.Close()
+
+		return
+	}
+
+	upstreamAddr := net.JoinHostPort(dstIP.String(), fmt.Sprintf("%d", dstPort))
+
+	logger.Debug(ctx, "Proxying sandbox egress IP filter connection",
+		zap.String("source_addr", sourceAddr),
+		zap.String("upstream_addr", upstreamAddr),
+	)
+
+	proxy(conn, upstreamAddr)
+}
+
+// proxy proxies the connection to the upstream address.
+func proxy(conn net.Conn, upstreamAddr string) {
+	dp := &tcpproxy.DialProxy{
+		Addr:        upstreamAddr,
+		DialTimeout: upstreamDialTimeout,
+	}
+	dp.HandleConn(conn)
+}
+
+// isEgressAllowed checks if egress is allowed based on domain and CIDR rules.
+// Priority order:
+//  1. Allow domain / Allow CIDR (if either matches → allow)
+//  2. Deny domain / Deny CIDR (if either matches → deny)
+//  3. Default: allow
+func isEgressAllowed(sbx *sandbox.Sandbox, hostname string, ip net.IP) (bool, error) {
```

Comment:
This function is really important, and we'd benefit from having tests covering it all.